subworkflows/local/bamfiltering.nf

```diff
@@ -0,0 +1,132 @@
+    // Generate BAM file of quality filtered and mapped-only reads,
+    // optionally retaining unmapped reads, defined in modules.config
+
+    SAMTOOLS_VIEW ( ch_bam_for_qualityfilter, [], [] ) // fasta isn't needed until we support CRAM
+    ch_versions = ch_versions.mix( SAMTOOLS_VIEW.out.versions.first() )
+
+    SAMTOOLS_FILTER_INDEX ( SAMTOOLS_VIEW.out.bam )
+    ch_versions = ch_versions.mix( SAMTOOLS_FILTER_INDEX.out.versions.first() )
+
+    ch_bam_for_genomics = SAMTOOLS_VIEW.out.bam.join( SAMTOOLS_FILTER_INDEX.out.bai )
+
+    // Only run if we actually remove mapped reads
+    if ( params.bamfiltering_mappingquality != 0 || params.bamfiltering_minreadlength != 0  ) {
+        SAMTOOLS_FLAGSTAT_FILTERED ( ch_bam_for_genomics )
+        ch_versions      = ch_versions.mix( SAMTOOLS_FLAGSTAT_FILTERED.out.versions.first() )
+        ch_multiqc_files = ch_multiqc_files.mix( SAMTOOLS_FLAGSTAT_FILTERED.out.flagstat )
+    }
+
+    //
+    // Metagenomics FASTQ generation for metagenomics (or just generation)
+    // - FASTQ generation is now separate from BAM filtering -
+    //    no length/quality filtering applies to metagenomic bam
+    //
+
+    // Generate unmapped bam (no additional filtering) if the unmapped bam OR unmapped for metagneomics selected
+    if ( params.bamfiltering_generateunmappedfastq || ( params.run_metagenomicscreening && params.metagenomicscreening_input == 'unmapped' ) ) {
+        SAMTOOLS_FASTQ_UNMAPPED ( bam.map{[ it[0], it[1] ]}, false )
+        ch_versions = ch_versions.mix( SAMTOOLS_FASTQ_UNMAPPED.out.versions.first() )
+    }
+
+    // Solution to the Andrades Valtueña-Light Problem: mapped bam for metagenomics (with options for quality- and length filtered)
+
+    if ( params.bamfiltering_generatemappedfastq ||  ( params.run_metagenomicscreening && ( params.metagenomicscreening_input == 'mapped' || params.metagenomicscreening_input == 'all' ) ) ) {
+        SAMTOOLS_FASTQ_MAPPED ( bam.map{[ it[0], it[1] ]}, false )
+        ch_versions = ch_versions.mix( SAMTOOLS_FASTQ_MAPPED.out.versions.first() )
+    }
+
+    if ( ( params.run_metagenomicscreening && params.metagenomicscreening_input == 'unmapped' ) && params.preprocessing_skippairmerging ) {
+        ch_paired_fastq_for_cat = SAMTOOLS_FASTQ_UNMAPPED.out.fastq
+                                    .mix(SAMTOOLS_FASTQ_UNMAPPED.out.singleton)
+                                    .mix(SAMTOOLS_FASTQ_UNMAPPED.out.other)
+                                    .groupTuple()
+                                    .map {
+                                        meta, fastqs ->
+                                            def meta_new = meta.clone()
+                                            meta_new['single_end'] = true
+                                        [ meta_new, fastqs.flatten() ]
+                                    }
+        CAT_FASTQ_UNMAPPED ( ch_paired_fastq_for_cat )
+    }
+
+    // TODO: see request https://github.com/nf-core/eager/issues/945
```

Comment:
Leftover TODO

Reply:
That I need to come back to later, it requires more complex thinking than my current capacities